hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/CustomContractCreationProcessor.java

```diff
@@ -16,20 +16,81 @@
+ * explicit {@code sender.decrementBalance(frame.getValue())} and
+ * {@code contract.incrementBalance(frame.getValue())} calls with a single call
+ * to the {@link HederaWorldUpdater#tryTransferFromContract(Address, Address, long, boolean)}
+ * dispatch method.
+ */
 public class CustomContractCreationProcessor extends ContractCreationProcessor {
+    // By convention, the halt reason should be INSUFFICIENT_GAS when the contract already exists
+    private static final Optional<ExceptionalHaltReason> COLLISION_HALT_REASON =
+            Optional.of(ExceptionalHaltReason.INSUFFICIENT_GAS);
+
     public CustomContractCreationProcessor(
             @NonNull final EVM evm,
             @NonNull final GasCalculator gasCalculator,
             final boolean requireCodeDepositToSucceed,
             @NonNull final List<ContractValidationRule> contractValidationRules,
             final long initialContractNonce) {
-        super(gasCalculator, evm, requireCodeDepositToSucceed, contractValidationRules, initialContractNonce);
+        super(
+                requireNonNull(gasCalculator),
+                requireNonNull(evm),
+                requireCodeDepositToSucceed,
+                requireNonNull(contractValidationRules),
+                initialContractNonce);
+    }
+
+    @Override
+    public void start(@NonNull final MessageFrame frame, @NonNull final OperationTracer tracer) {
+        final var addressToCreate = frame.getContractAddress();
+        final var contract =
+                frame.getWorldUpdater().getOrCreate(addressToCreate).getMutable();
+
+        if (alreadyCreated(contract)) {
+            halt(frame, COLLISION_HALT_REASON);
+            tracer.traceAccountCreationResult(frame, COLLISION_HALT_REASON);
+        } else {
+            final var updater = proxyUpdaterFor(frame);
+            // A contract creation is never a delegate call, hence the false argument below
+            final var maybeReasonToHalt = updater.tryTransferFromContract(
+                    frame.getSenderAddress(), addressToCreate, frame.getValue().toLong(), false);
+            if (maybeReasonToHalt.isPresent()) {
+                // Besu doesn't trace the creation on a modification exception, so seems
+                // like we shouldn't do it here either; but may need a bit more consideration
+                halt(frame, maybeReasonToHalt);
+            } else {
+                contract.setNonce(INITIAL_CONTRACT_NONCE);
+                frame.setState(MessageFrame.State.CODE_EXECUTING);
+            }
+        }
+    }
+
+    private void halt(final MessageFrame frame, @NonNull final Optional<ExceptionalHaltReason> reason) {
+        frame.setState(MessageFrame.State.EXCEPTIONAL_HALT);
+        frame.setExceptionalHaltReason(reason);
+    }
+
+    private boolean alreadyCreated(final MutableAccount account) {
+        return account.getNonce() > 0 || account.getCode().size() > 0;
```

Comment:
Are there contracts with code size == 0 that nevertheless have nonce > 0 (they've been called)?  Just curious.

Reply:
The [initial contract nonce](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/processors/ProcessorModule.java#L29) is `1`, so _if_ it's possible to create a contract whose bytecode is empty, then we would have the case you describe.

I'm not sure if it's possible to create a contract with empty bytecode tho. 🤔